### PR TITLE
Fk remove zarr ss2 pipeline GL-1007

### PR DESCRIPTION
**Pull request by khajoue2**
### Purpose
Zarr files are removed from Optimpus pipeline, ss2 pipeline single sample and multi sample

### Changes
Loom files are produced instead of zarr
- No changes.

### Review Instructions

- No instructions.

### CODE REVIEW

File: pipelines/optimus/Optimus.changelog.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+#optimus_v2.1.0
+
+* Removed zarr file
+
+* Added emptyDrops for sn_rna mode
```
Comment (kbergin) [inline pipelines/optimus/Optimus.changelog.md:5]: Didn't we actually remove emptyDrops, not add it?

I think we need to add a bit more information about the removal of zarr.

Something like "Removed zarr formatted matrix and metrics outputs and replaced with loom" and I believe this is a breaking change. If this is going out with the features for the snRNAseq support it's definitely a major version update. Are the snRNAseq changes included in this PR?

Comment (kbergin) [reply pipelines/optimus/Optimus.changelog.md:5]: There was a note in the group review to mention in changelog (and all changelogs here) that we added Gene Name as an attribute in the matrix.

File: library/tasks/LoomUtils.wdl [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  runtime {
+    docker: docker
+    cpu: 4  # note that only 1 thread is supported by pseudobam
+    memory: "18 GiB"
+    disks: "local-disk 100 HDD"
+    preemptible: preemptible
+  }
+
+  output {
+    File loom_output = "output.loom"
```
Comment (kbergin) [inline library/tasks/LoomUtils.wdl:117]: Can we name this something more descriptive? Perhaps use the sample id that's input to the pipeline as the basename?

File: docker/loom-output/create_loom_ss2.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    reader = csv.DictReader(open(rsem_gene_results_path), delimiter="\t")
+
+    expression_values = {}
+    count_values = {}
+    for row in reader:
+        expression_values[row["gene_id"]] = float(row["TPM"])
+        count_values[row["gene_id"]] = float(row["expected_count"])
+
+    sorted_gene_ids = sorted(expression_values.keys())
+    sorted_tpms = [expression_values[g] for g in sorted_gene_ids]
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply docker/loom-output/create_loom_ss2.py]: Added!

File: pipelines/optimus/Optimus.changelog.md (hunk 1 above)
Comment (khajoue2) [reply pipelines/optimus/Optimus.changelog.md:5]: Done!

File: library/tasks/LoomUtils.wdl [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      
+      # Merge the loom files
+      python3 /tools/ss2_loom_merge.py --input-loom-files ${sep=' ' loom_input} \
+      --output-loom-file output.loom --plate-sample-id plateid
+
+
+    }
+
+    output {
+        File loom_output_file = "output.loom"
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply library/tasks/LoomUtils.wdl:145]: Changed it to sample_id and plateid

File: library/tasks/LoomUtils.wdl [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+       --sample_id ${sample_id} \
+       --count_matrix ${sparse_count_matrix} \
+       --expression_data_type $EXPRESSION_DATA_TYPE_PARAM
+  }
+
+  runtime {
+    docker: docker
+    cpu: 4  # note that only 1 thread is supported by pseudobam
+    memory: "18 GiB"
+    disks: "local-disk 100 HDD"
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply library/tasks/LoomUtils.wdl:112]: We need to make a ticket for this :)

File: docker/loom-output/create_loom_optimus.py [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+        else:
+            # if no metrics for a cell present in the count matrix then fill them with np.nans
+            gene_metric_values.append([np.nan] * ncols)
+
+    nrows = len(gene_ids)
+    
+    #Generate row attribute dictionary
+    row_attrs = { "gene_names":gene_names, "ensembl_ids": gene_ids}
+
+    X=np.array(gene_metric_values)
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply docker/loom-output/create_loom_optimus.py]: Done!

File: docker/loom-output/create_loom_optimus.py [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+    if add_emptydrops_results == 'yes':
+       emptydrops_df = pd.read_csv(args.empty_drops_file, dtype=str)
+
+    # Check that input is valid
+    if metrics_df.shape[0] == 0 or metrics_df.shape[1] == 0:
+        logging.error("Cell metrics table is not valid")
+        raise ValueError()
+
+    if add_emptydrops_results == 'yes':
+        if emptydrops_df.shape[0] == 0 or emptydrops_df.shape[1] == 0:
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply docker/loom-output/create_loom_optimus.py]: Done!

File: docker/loom-output/create_loom_optimus.py [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+       
+    else:
+        final_df = cellorder_df.merge(metrics_df, on="cell_id", how="left")
+        ColumnNames = IntColumnNames + FloatColumnNames 
+        BoolColumnNames = []
+
+    # Split the dataframe
+    final_df_float = final_df[ColumnNames]
+
+    # Data types for storage
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply docker/loom-output/create_loom_optimus.py]: Done!

File: pipelines/optimus/Optimus.changelog.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (khajoue2) [reply pipelines/optimus/Optimus.changelog.md:5]: Done

File: pipelines/optimus/Optimus.wdl [hunk 9]
```diff
@@ -61,7 +58,7 @@ workflow Optimus {
   }
 
   # version of this pipeline
-  String version = "optimus_v2.0.0"
+  String version = "optimus_v2.1.0"
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply pipelines/optimus/Optimus.wdl]: Done!

File: docker/loom-output/create_loom_optimus.py [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+        name = numeric_field_names[i]
+        data = X[:, i]
+        row_attrs[name] = data
+    if args.verbose:
+        logging.info("# of genes: {0}".format(nrows))
+        logging.info("# of gene metadate metrics: {0}".format(ncols))
+    return row_attrs
+
+def generate_col_attr(args):
+    """Converts cell metrics from the Optimus pipeline to zarr file
```
Comment (ekiernan) [inline docker/loom-output/create_loom_optimus.py:123]: Do you need to remove this line about the zarr file?

Comment (khajoue2) [reply docker/loom-output/create_loom_optimus.py:123]: Changed it :)

File: docker/loom-output/create_loom_optimus.py [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+            [ "emptydrops_LogProb", "emptydrops_PValue", "emptydrops_FDR" ]
+        BoolColumnNames = ["emptydrops_Limited", "emptydrops_IsCell"]
+       
+    else:
+        final_df = cellorder_df.merge(metrics_df, on="cell_id", how="left")
+        ColumnNames = IntColumnNames + FloatColumnNames 
+        BoolColumnNames = []
+
+    # Split the dataframe
+    final_df_float = final_df[ColumnNames]
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply docker/loom-output/create_loom_optimus.py:230]: Done.

File: docker/loom-output/create_loom_optimus.py [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+    if args.verbose:
+        logging.info(
+            "Added cell metadata_bool with {0} rows and {1} columns".format(
+                final_df_bool.shape[0], final_df_bool.shape[1]
+            )
+        )
+
+    return col_attrs
+
+def generate_matrix(args):
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply docker/loom-output/create_loom_optimus.py:282]: We decided to make this changes after the next release.

File: pipelines/optimus/Optimus.changelog.md [hunk 13]
```diff
@@ -1,3 +1,9 @@
+#optimus_v2.1.0
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply pipelines/optimus/Optimus.changelog.md]: Done!

File: pipelines/smartseq2_single_sample/SmartSeq2SingleSample.changelog.md [hunk 14]
```diff
@@ -1,3 +1,9 @@
+# smartseq2_v4.0.0
+
+2020-06-04 (Date of Last Commit)
+
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply pipelines/smartseq2_single_sample/SmartSeq2SingleSample.changelog.md:4]: Done

File: pipelines/smartseq2_multisample/MultiSampleSmartSeq2.changelog.md [hunk 15]
```diff
@@ -1,3 +1,8 @@
+# MultiSampleSmartSeq2_v2.0.0
+
+2019-06-04 (Date of Last Commit)
+* Removed zarr output and made loom output as default
+
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply pipelines/smartseq2_multisample/MultiSampleSmartSeq2.changelog.md:5]: Done.

File: docker/loom-output/create_loom_optimus.py (hunk 10 above)
Comment (khajoue2) [reply docker/loom-output/create_loom_optimus.py:123]: Changed "zarr" to "loom"

File: library/tasks/LoomUtils.wdl (hunk 2 above)
Comment (khajoue2) [reply library/tasks/LoomUtils.wdl]: Done.

File: docker/loom-output/create_loom_optimus.py (hunk 10 above)
Comment (jessicaway) [reply docker/loom-output/create_loom_optimus.py:123]: Still see "zarr" here

File: docker/loom-output/create_loom_optimus.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jessicaway) [reply docker/loom-output/create_loom_optimus.py:282]: Can we write this in a ticket for the backlog?

File: pipelines/smartseq2_single_sample/SmartSeq2SingleSample.changelog.md [hunk 16]
```diff
@@ -1,3 +1,13 @@
...
+* Added loom output and removed the zarr output
+
+* Loom file attribute names have changed: CellID: cell_names, Gene: gene_names and Accession: ensembl_ids
+
+* Loom file name has changed from out.loom to "sample_id".loom
+
```
Comment (jessicaway) [inline pipelines/smartseq2_single_sample/SmartSeq2SingleSample.changelog.md:10]: We should include the addition of the expected counts in the loom output here

File: pipelines/smartseq2_multisample/MultiSampleSmartSeq2.changelog.md [hunk 17]
```diff
@@ -1,3 +1,12 @@
...
+* Loom file attribute names have changed: CellID: cell_names, Gene: gene_names and Accession: ensembl_ids
+
+* Loom file name has changed from out.loom to "plateid".loom
+
```
Comment (jessicaway) [inline pipelines/smartseq2_multisample/MultiSampleSmartSeq2.changelog.md:9]: Add the expected counts output in loom to the changelog please

Comment (kbergin) [reply pipelines/smartseq2_multisample/MultiSampleSmartSeq2.changelog.md:9]: It looks like it was in the previous release on 5/7 changelog, but we should update it from estimated -> expected counts to be consistent?

Comment (jessicaway) [reply pipelines/smartseq2_multisample/MultiSampleSmartSeq2.changelog.md:9]: Good point!

File: docker/loom-output/create_loom_optimus.py (hunk 10 above)
Comment (khajoue2) [reply docker/loom-output/create_loom_optimus.py]: Pushed new changes.

File: pipelines/smartseq2_single_sample/SmartSeq2SingleSample.changelog.md (hunk 16 above)
Comment (khajoue2) [reply pipelines/smartseq2_single_sample/SmartSeq2SingleSample.changelog.md:10]: Done.

File: pipelines/smartseq2_multisample/MultiSampleSmartSeq2.changelog.md (hunk 17 above)
Comment (khajoue2) [reply pipelines/smartseq2_multisample/MultiSampleSmartSeq2.changelog.md:9]: Done

File: docker/loom-output/loomCompare.py [hunk 18]
```diff
@@ -17,8 +17,8 @@ def main():
     truth_loom = loompy.connect(args.truth_loom_path)
     check_loom = loompy.connect(args.check_loom_path)
 
-    truth_loom_array = pd.DataFrame(data=truth_loom[:, :], index=truth_loom.row_attrs['Gene'], columns=truth_loom.col_attrs['CellID'])
-    check_loom_array = pd.DataFrame(data=check_loom[:,:], index=check_loom.row_attrs['Gene'], columns=check_loom.col_attrs['CellID'])
+    truth_loom_array = pd.DataFrame(data=truth_loom[:, :], index=truth_loom.row_attrs['gene_names'], columns=truth_loom.col_attrs['cell_names'])
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply docker/loom-output/loomCompare.py:20]: We should discuss this in more details.

**Pull request merged**